posthog/caching/insight_caching_state.py

```diff
@@ -0,0 +1,136 @@
+from datetime import timedelta
+from enum import Enum
+from functools import cached_property
+from typing import Optional, Union
+
+from django.utils.timezone import now
+
+from posthog.caching.utils import active_teams
+from posthog.models.dashboard_tile import DashboardTile
+from posthog.models.insight import Insight, InsightViewed, generate_insight_cache_key
+from posthog.models.insight_caching_state import InsightCachingState
+from posthog.models.team import Team
+
+VERY_RECENTLY_VIEWED_THRESHOLD = timedelta(hours=48)
+GENERALLY_VIEWED_THRESHOLD = timedelta(weeks=2)
+MAX_ATTEMPTS = 3
+
+
+# :TODO: Make these configurable
+class TargetCacheAge(Enum):
+    NO_CACHING = None
+    LOW_PRIORITY = timedelta(days=7)
+    MID_PRIORITY = timedelta(hours=24)
+    HIGH_PRIORITY = timedelta(hours=12)
+
+
+# Helps do large-scale re-calculations efficiently by loading some data only once
+class LazyLoader:
+    @cached_property
+    def active_teams(self):
+        return active_teams()
+
+    @cached_property
+    def recently_viewed_insights(self):
+        recently_viewed_insights = InsightViewed.objects.filter(
+            last_viewed_at__gte=now() - VERY_RECENTLY_VIEWED_THRESHOLD
+        ).distinct("insight_id")
+        return set(recently_viewed_insights.values_list("insight_id", flat=True))
+
+
+def upsert(
+    team: Team, target: Union[DashboardTile, Insight], lazy_loader: Optional[LazyLoader] = None
+) -> Optional[InsightCachingState]:
+    lazy_loader = lazy_loader or LazyLoader()
+    cache_key = calculate_cache_key(team, target)
+    if cache_key is None:  # Non-cachable model
+        return None
+
+    target_age = calculate_target_age(team, target, lazy_loader)
+    target_age_seconds = target_age.value.total_seconds() if target_age.value is not None else None
+    caching_state, _ = InsightCachingState.objects.update_or_create(
+        team_id=team.pk,
+        insight=target if isinstance(target, Insight) else target.insight,
+        dashboard_tile=target if isinstance(target, DashboardTile) else None,
+        defaults={
+            "cache_key": cache_key,
+            "target_cache_age_seconds": target_age_seconds,
+        },
+    )
+    return caching_state
+
+
+def fetch_states_in_need_of_updating():
+    return InsightCachingState.objects.raw(
```

Comment:
I think we can/should do this within the ORM to stop it accruing complexity (cf the property definition API query)

I think it needs an annotation to get the right-hand side of the `last_refresh` filter

Reply:
Tried it with ORM, we need to start doing really janky things to get the `target_cache_age` filter to work that will just break down over time.

Given we already write a bunch of SQL, I'd rather maintain this than that jank :)